orca-keel/src/test/java/com/netflix/spinnaker/orca/keel/task/ImportDeliveryConfigTaskTest.java

```diff
@@ -0,0 +1,251 @@
+  }
+
+  /**
+   * This test is a positive case which verifies if the task returns {@link
+   * ImportDeliveryConfigTask.SpringHttpError} on 4xx http error. Here the error body is mocked with
+   * timestamp in {@link ChronoUnit#MILLIS}, which will be parsed to exact same timestamp in the
+   * method @see {@link ImportDeliveryConfigTask#handleRetryableFailures(RetrofitError,
+   * ImportDeliveryConfigTask.ImportDeliveryConfigContext)} and results in successful assertions of
+   * all the fields.
+   *
+   * <p>Other positive scenarios when the timestamp results in accurate value, are when the units in
+   * : {@link ChronoUnit#SECONDS} {@link ChronoUnit#DAYS} {@link ChronoUnit#HOURS} {@link
+   * ChronoUnit#HALF_DAYS} {@link ChronoUnit#MINUTES}
+   */
+  @Test
+  public void verifyTaskReturnsSpringHttpError() throws JsonProcessingException {
+
+    var mockResponseBody =
+        Map.of(
+            "name",
+            "keeldemo-manifest",
+            "application",
+            "keeldemo",
+            "artifacts",
+            Collections.emptySet(),
+            "environments",
+            Collections.emptySet());
+
+    var httpStatus = HttpStatus.BAD_REQUEST;
+
+    // Initialize SpringHttpError with timestamp in milliseconds and HttpStatus 400 bad request.
+    var httpError = mockSpringHttpError(httpStatus, Instant.now().truncatedTo(ChronoUnit.MILLIS));
+
+    var context = Map.of("error", httpError);
+
+    TaskResult terminal = TaskResult.builder(ExecutionStatus.TERMINAL).context(context).build();
+
+    // simulate 400 bad request http error with SpringHttpError body
+    simulateFault("/delivery-configs/", objectMapper.writeValueAsString(httpError), httpStatus);
+
+    when(scmService.getDeliveryConfigManifest(
+            (String) contextMap.get("repoType"),
+            (String) contextMap.get("projectKey"),
+            (String) contextMap.get("repositorySlug"),
+            (String) contextMap.get("directory"),
+            (String) contextMap.get("manifest"),
+            (String) contextMap.get("ref")))
+        .thenReturn(mockResponseBody);
+
+    var result = importDeliveryConfigTask.execute(stage);
+
+    assertThat(result).isEqualTo(terminal);
+  }
+
+  /**
+   * This test is a negative case which verifies if the task returns {@link
+   * ImportDeliveryConfigTask.SpringHttpError} on 4xx http error. Here the error body is mocked with
+   * timestamp in {@link ChronoUnit#NANOS}, which WILL NOT be parsed to exact timestamp in the
+   * method @see {@link ImportDeliveryConfigTask#handleRetryableFailures(RetrofitError,
+   * ImportDeliveryConfigTask.ImportDeliveryConfigContext)} and results in will contain incorrect
+   * timestamp.
+   *
+   * <p>Other cases where the timestamp will result in incorrect value, are when the units in :
+   * {@link ChronoUnit#MICROS}
+   */
+  @Test
+  public void verifyTaskResultReturnsSpringHttpErrorWhenTimestampIsInNanos()
```

Comment:
There's a fair amount of duplication here. Can you parameterize this test across the various ChronoUnits you're testing?

Reply:
I'd also like to see a test where the response from keel includes no timestamp at all.

Reply:
Writing a test on this would result in NullPointerException, and the test will not get executed.

Reply:
Or maybe a better thing to ask is, where does the NullPointerException come from?

Reply:
Here is how the SpringHttpError class is defined

```
data class SpringHttpError(
    val error: String,
    val status: Int,
    val message: String? = error,
    val timestamp: Instant = Instant.now(),
    val details: Map<String, Any?>? = null // this is keel-specific
  )
```
  The parameter without '?' symbol are @NotNull annotated and the parameters with '?' symbol are @Nullable .
  
  So , when we pass null values to the @NotNull parameters , the  NullPointerException is thrown at the init stage itself. 
  
  Here is the exact exception thrown : 
  
 ` java.lang.NullPointerException: Parameter specified as non-null is null: method com.netflix.spinnaker.orca.keel.task.ImportDeliveryConfigTask$SpringHttpError.<init>, parameter timestamp`

Reply:
So, change the code to check for null before calling the constructor.

Reply:
Logic added to handle the case when the timestamp in SpringHttpError is null. The constructor will initialize  this field with the default value "Instant.now()"/ current time.